src/vos/vos_io.c

```diff
@@ -1104,14 +1105,44 @@ akey_fetch_recx(daos_handle_t toh, const daos_epoch_range_t *epr,
 			holes += lo - index;
 		}
 
-		/* Hole extent, with_shadow case only used for EC obj */
-		if (bio_addr_is_hole(&ent->en_addr) ||
-		    (with_shadow && (ent->en_epoch < shadow_ep))) {
+		if ((with_shadow && (ent->en_epoch < shadow_ep))) {
```

Comment:
If `shadow_ep` is newer, is it still hole? As my understand, under such case, current `ent` should be ignored since `shadow` will cover this `ent`, right?

So here, you may want express: `ent->en_epoch > shadow_ep` ?

On the other hand, we also need to consider the minor epoch if the major epoch are the same.

Reply:
as discussed offline, need not consider minor epoch now. thanks.

Reply:
Discussed with Xuezhao, the logic can work.